File: strings/Knuth_Morris_Pratt.py
"""
Date: 06-15-24

Knuth Morris Pratt


This algorithm is used to find occurrences of a pattern in a text.

Complexity: O(n+m)
  n: Length of text
  m: Length of pattern

Used propierty:
  - Degenerating property: Pattern having the same sub-patterns appearing more than
    onde in the pattern.

How does it works?
The basic idea behaind this algorithm is to skip characters in text using the sub-patterns
found in the pattern.

This algorithm is divided in 2 main steps:
1. Compute the lps array (Find the sub-pattern and therefore knowing how many characters be can skip)
2. Use lps array in the search to skip those characters that are not necessary to evaluate.

https://www.geeksforgeeks.org/kmp-algorithm-for-pattern-searching/
"""


"""
  This function pre-processes the pattern before of comparing with the string.
  Thanks to the lps[] array, which contains the number of letter to be skipped, is that 
  the match pattern can be optimized, avoiding revalidation multiple cases.
"""
def prefix_function(pattern):
  # Creating lps[] array
  # Note: lps comes from longest prefix suffix

  # Creating variables needed for the algorithm
  lps = [0] * len(pattern) # Array to store the calculations
  lenPrefix = 0 # Array to store the length of the patterns found
  # Iterator. It starts is initialized with 1, because "0", which is the first letter, has 
  # already covered
  i = 1 

  # Initializing lps. Since we haven't begun to analyze the pattern, it is always going to start in 
  # 0. 
  lps[lenPrefix] = 0

  while(i < len(pattern)):
    if pattern[lenPrefix] == pattern[i]:
      # That means that there is a match in the prefix (lenPrefix) and the suffix (i)
      lenPrefix += 1
      lps[i] = lenPrefix
      i += 1
    else:
      if lenPrefix > 0:
        # That means that there was not a match between the prefix and the suffix.
        # So, lenPrefix is not 0 (there are other letter in the prefix), there is the
        # Possibility
        lenPrefix = lps[lenPrefix - 1]
      else:
        lenPrefix = 0
        lps[i] = lenPrefix
        i += 1

  return lps


"""
Once you have the lps arrays, you only have to make an linear search comparing the current letter
of the text with the current letter of the pattern.

The difference is when you detect that there is no match between both iterator, you have just to use
lps to know how many characters you can skip.

Remember for each character in the pattern, there is an value in lps indicating the number of 
characters to skip.
"""
def find_match(text, lps, pattern):
  i = 0
  j = 0

  while(i < len(text)):
    if text[i] == pattern[j]:
      i += 1
      j += 1
    else:
      i += lps[j]
      j = 0

    if j == len(pattern):
      print(text[i - len(pattern):i])
      print("Pattern found")
      j = 0


def main():
  lps = prefix_function("ABABC")
  find_match("ABABABABCAAA", lps, "ABABC")

main()

File: strings/Knuth_Morris_Pratt.cpp
/*
  Author: Renet de Jesús Pérez Gómez
  Date: 01-30-24

  Algorithm to implement: Knuth Morris Pratt (KMP)

  This algorithm finds all the occurrences of a pattern in a text.

  Unlike the naive method, on which you need to compare all the possibilities for the pattern, KMP
  use the "degenerating propierty".

  According to geeksforgeeks degenerating propierty means:
  "(pattern having the same sub-patterns appearing more than once in the pattern)"  
*/

#include <iostream>
#include <vector>
#include <string>

using namespace std; 

class KMP {
    private:
    vector<string> pat;
    vector<int> lps;
    vector<string> text;

    void computeLPSArray() {
      int len = 0;
      
      for(int i = 0; i < this->pat.size(); i++) {
        if(i == 0) {
          // lps[0] always is 0.
          lps.push_back(0);
        } else if(pat[len] == pat[i]) {
          /*
            Both characters are equal.
            
            Explanation:
              Pattern having the same sub-patterns appearing more than once in a pattern
          */
          len += 1;
          lps.push_back(len); 
        } else {
          /*
            It is needed to find another charater that match with the current character,
            this looking in previous values in the array.

            There is 2 cases: There is not a match then len = 0 or there is a match; therefore 0 = n 
          */
          bool findMatch = true;
          while(findMatch) {
            if(pat[len] == pat[i]) {
              len += 1;
              findMatch = false;              
            } else {
              len -= 1;
              if(len > 0) {
                /* Do Nothing */
              } else {
                len = 0;
                findMatch = false;
              }
            }
          }
          lps.push_back(len);
        }
      }
    }

  public:
    KMP(string& pattern, string& text) {      
      for(const auto &element : pattern) {
        this->pat.push_back(string(1, element));
      }

      for(const auto &element : text) {
        this->text.push_back(string(1, element));
      }
    } 

    void solvePattern() {
      int i = 0;
      int j = 0;
      int textLength = this->text.size();
      int pathLenght = this->pat.size();

      this->computeLPSArray();

      // Find pattern in text
      while(i < textLength) {
        if(this->text[i] == this->pat[j]) {
          /*
            It was found a coincidence between the current character in both the pattern 
            and in the text, so we move forward in the iteration.
          */
          i += 1;
          j += 1;
        } else {
          if(j > 0) {
            /*
              That means that the current position is not a prefix for the next coincidence, but it might be
              the previous prefix in lps vector.
            */
            j = lps[j - 1];
          } else {
            /*
              That means that it was not found any match for the current character and there is not a
              useful prefix in lps. 
            */
            i += 1;
          }
        }

        if(j == pathLenght) {
          /*
            It was found a match for the whole pattern.
            
            As it was a match for the entire pattern, j needs to be reset,
            in the new assignation it is only substracted 1, because the current
            suffix can be the prefix of the next pattern.
          */
          cout << "Pattern  found at index: " <<  i - pathLenght << endl;
          
          j = lps[j - 1]; 
        } 
      }
    }
};

int main(){

  // string text = "ABABDABACDABABCABAB";
  // string pattern = "ABABCABAB";

  string text = "AAAAABAAABA";
  string pattern = "AAAA";
  KMP instance(pattern, text);
  instance.solvePattern();
  return 0;
}